### PR TITLE
Added a very basic neural CDE model

### PR DESCRIPTION
**Pull request by packquickly**
Added a very simple neural CDE. It is not particularly extensible or powerful, but a fine basic example.

Sets roughly the structure I'd like to use use in the repo, namely `equinox_monorepo/model_name/*` where * is all the files for the model, including a short README about the model.

### CODE REVIEW

File: equinox_monorepo/neural_cde/neural_cde.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+from typing_extensions import TypeAlias
+
+# These type aliases are unnecessary, but I (packquickly) like them :)
+Linear: TypeAlias = eqx.nn.Linear
+MLP: TypeAlias = eqx.nn.MLP
+Interpolator: TypeAlias = dfx.AbstractGlobalInterpolation
+Solver: TypeAlias = dfx.AbstractSolver
+
+
+class Mode(Enum):
```
Comment (Benjamin-Walker) [inline equinox_monorepo/neural_cde/neural_cde.py]: I'm not sure I see the reason for using an Enum to specify whether the model takes the terminal value or meanpool value, is there a reason for this?

File: equinox_monorepo/neural_cde/neural_cde.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+        depth: int,
+        activation: Callable,
+        final_activation: Callable,
+        *,
+        key: PRNGKeyArray,
+        **kwargs,
+    ):
+        """**Arguments:**
+
+
```
Comment (Benjamin-Walker) [inline equinox_monorepo/neural_cde/neural_cde.py]: Unnecessary white space

File: equinox_monorepo/neural_cde/neural_cde.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+            feedforward network output.
+        - `hidden_width`: The hidden size of the feedforward network, including the
+            output layer.
+        - `depth`: The depth of the feedforward network.
+        - `activation`: The activation between each hidden layer.
+        - `final_activation`: The activation function after the output layer.
+        - `key`: A `jax.random.PRNGKey` used to provide randomness for parameter
+            initialisation.
+        """
+        super().__init__(**kwargs)
```
Comment (Benjamin-Walker) [inline equinox_monorepo/neural_cde/neural_cde.py]: I think we can remove the `**kwargs` and `super().__init__(**kwargs)` as I don't think we will be passing anything to eqx.Module

File: equinox_monorepo/neural_cde/neural_cde.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        self.data_size = data_size
+        self.state_size = model_size
+
+        mlp = eqx.nn.MLP(
+            in_size=self.state_size,
+            out_size=model_size * data_size,
```
Comment (Benjamin-Walker) [inline equinox_monorepo/neural_cde/neural_cde.py]: Change `model_size * data_size` to `self.state_size * self.data_size` to be consistent with using self variables once they have been declared (or change `in_size=self.state_size` to `in_size=model_size`.

File: equinox_monorepo/neural_cde/neural_cde.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            Defaults to `diffrax.LinearInterpolation.`
+        - `solver`: The Diffrax solver to use to solve the neural CDE. Defaults to
+            `diffrax.Tsit5()`.
+        - `mode`: Determines how to go from an output of shape (times, feats) to
+            (feats) for use in classification. Uses the enum `Mode`. Defaults to
+            `Mode.terminal`.
+        - `key`: A `jax.random.PRNGKey` used to provide randomness for parameter
+            initialisation.
+        """
+        super().__init__(**kwargs)
```
Comment (Benjamin-Walker) [inline equinox_monorepo/neural_cde/neural_cde.py]: Same as above

File: equinox_monorepo/neural_cde/neural_cde.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            out = solution.ys[-1]
+        elif self.mode == Mode.meanpool:
+            out = jnp.mean(solution.ys, axis=0)
+        else:
+            raise ValueError(
+                "'mode' should be either `Mode.terminal` or "
+                f"`Mode.meanpool` using the `Mode` enum type. Got {self.mode} "
+            )
+
+        return jnn.softmax(self.decoder(out))
```
Comment (Benjamin-Walker) [inline equinox_monorepo/neural_cde/neural_cde.py]: If self.decoder has output dimension of 1 I think using softmax as the final activation will fix the output to always be 1, could either have binary classification's changed to have output dimension of 2 or use a different final activation when n_classes=2

**Review by Benjamin-Walker — COMMENTED** [review]
Mostly looks good, left a few minor comments

File: equinox_monorepo/neural_cde/neural_cde.py (hunk 1 above)
Comment (packquickly) [reply equinox_monorepo/neural_cde/neural_cde.py]: The model only offers a few options ("terminal", "meanpool") which the user may want to switch between, so in this case I prefer enums over something like a string (better type safety, avoid misspellings, are easier to refactor, self-documenting, etc.)

Comment (Benjamin-Walker) [reply equinox_monorepo/neural_cde/neural_cde.py]: I've not used an Enum as a class attribute before, but my understanding is that this would require you to import Mode from this file alongside NCDEClassifier in order to swap to meanpool? I feel this is not as clean as being able to import just NCDEClassifier and specify all of its attributes.

Comment (Benjamin-Walker) [reply equinox_monorepo/neural_cde/neural_cde.py]: If the above is the case, but you would still prefer to use an Enum, then I think that the name should be made more specific, like `PoolingModeNCDE`, so its use is clearer when imported into other scripts

Comment (packquickly) [reply equinox_monorepo/neural_cde/neural_cde.py]: You're right, I think enums as class attributes in general feel "unequinox." Decided to simply remove it in favor of a bool switching between meanpool and terminal, since these two approaches dominate and this is a simple model already. An alternative would be to pass a callable, which gives the user more flexibility but is too complex for this model in my opinion.
